### PR TITLE
fix mac os memory leak

### PR DESCRIPTION
**Pull request by Lurk**
TLDR;
I did not find an easy way to clean up NSArray, but there is a better option to get NSString from pasteboard - stringForType.

Next up. Even if send release to NSString, there are leftovers after conversion NSString to Rust string. Right now, conversion happens in INSString.as_str(), which internally uses NSString.UTF8String. UTF8String, according to Apple docs, is a pointer lifetime of which is less or equal NSString. This is not true if the clipboard contains characters outside the ASCII range. That's why nsstring_to_rust_string function.

I tried to describe the full process in a [blogpost](https://barhamon.com/post/rust_and_nsstring)

### CODE REVIEW

**Comment by chrisduerr** [comment]
Please be more specific on where the memory leak is, how to reproduce it and why your changes are necessary to fix it.

**Comment by Lurk** [comment]
@chrisduerr Woah, that was fast responce :)

To reproduce memory leak on macOS

```rust
use copypasta::{ClipboardContext, ClipboardProvider};

fn main() {
    let mut ctx: ClipboardContext = ClipboardContext::new().unwrap();
    for _ in 0..20_000 {
        match ctx.get_contents() {
            Ok(str) => println!("{}", str.chars().count()),
            Err(_e) => println!("{:?}", _e),
        }
    }
}
```
```bash
cargo instruments -t Allocation
```
will produce a nice graph
full_glory

The leak occurs in at least two places. 
First, and I am not sure about the actual reason here, NSArray of NSString we got [here](https://github.com/alacritty/copypasta/blob/master/src/osx_clipboard.rs#L54)  does not get released.  If we send release to NSArray nothing happens but, if we send the release to NSString we got [here](https://github.com/alacritty/copypasta/blob/master/src/osx_clipboard.rs#L64) 
```rust
string_array[0]
```
memory leak drops ~3x

Since we use  NSArray to get the first NSString out of it, we can use [NSPasteboard::stringForType](https://developer.apple.com/documentation/appkit/nspasteboard/1533566-stringfortype?language=objc) which will return NSString. 

Now we can [send release directly to NSString](https://github.com/Lurk/copypasta/blob/master/src/osx_clipboard.rs#L56) we are getting from NSPastebord. 


now If we run 
 ```bash
cargo instruments -t Allocation
```
again, we will see this 
disappointment

Which is a huge 3x win but still. 

Next leak, again I am not sure that I identified the source precisely, is that we use INSString.to_str() which internally uses NSString.UTF8String. [Apple doc](https://developer.apple.com/documentation/foundation/nsstring/1411189-utf8string?language=objc) says

> This C string is a pointer to a structure inside the string object, which may have a lifetime shorter than the string object and will certainly not have a longer lifetime. Therefore, you should copy the C string if it needs to be stored outside of the memory context in which you use this property.

And most of the time it is true, I mean this C string gets cleaned up, except when clipboard contains characters outside the ASCII range. In that case it lives forever. 

Solution I come up with:
```rust
fn nsstring_to_rust_string(nsstring: *mut NSString) -> Result<String> {
    unsafe {
        let string_size: usize =
            msg_send![nsstring, lengthOfBytesUsingEncoding: NSUTF8StringEncoding];
        //we need +1 because getCString will return null terminated string
        let char_ptr = libc::malloc(string_size + 1);
        let res: bool = msg_send![nsstring, getCString:char_ptr  maxLength:string_size + 1 encoding:NSUTF8StringEncoding];
        if res {
            let c_string = CStr::from_ptr(char_ptr as *const i8);
            libc::free(char_ptr);
            Ok(c_string.to_string_lossy().into_owned())
        } else {
            libc::free(char_ptr);
            Err("Casting from NSString to Rust string has failed".into())
        }
    }
}
```
Now allocations is showing
final_result

I will gladly provide further explanations and/or fixes if needed.

**Comment by chrisduerr** [comment]
These seem like more fundamental issues with the `objc` crates. I find it difficult to believe that your solution is the 'correct' approach here. I try to stay away from macOS as much as possible, but surely it's possible to use these types without leaking.

I'd maybe try and look at some other usage of the macos crates in winit maybe. But manual allocation with libc and malloc should surely be unnecessary.

**Comment by Lurk** [comment]
Yeah, I discovered that leak in my first Rust project. Imagine how surprised I was :)

Anyway. I do not insist on this solution and will gladly use any other working solution. But every other clipboard crate I found in one way or another uses Avraham Weinstock implementation. And I looked hard, spent almost a month of evenings on it.

**Comment by chrisduerr** [comment]
Well if you're willing to look to further look into this, feel free. But the current solution does not seem appropriate to me.

**Comment by Lurk** [comment]
I am still trying to fix this because it annoys me as hell :)

Progress so far.

Found a way to get rid of libc.
Created repo with the reproducible example, fix, and benchmark (https://github.com/Lurk/nsstring_leak).
Created the issue in rust-objc-foundation (https://github.com/SSheldon/rust-objc-foundation/issues/15)

I was kind of hoping that after the macOS update leak will go away, but no luck.
